File: app/src/main/java/com/teambrella/android/ui/claim/ClaimActivity.java
package com.teambrella.android.ui.claim;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.gson.JsonObject;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.api.server.TeambrellaServer;
import com.teambrella.android.api.server.TeambrellaUris;
import com.teambrella.android.data.base.TeambrellaDataFragment;
import com.teambrella.android.data.base.TeambrellaDataPagerFragment;
import com.teambrella.android.image.TeambrellaImageLoader;
import com.teambrella.android.ui.base.ADataProgressFragment;
import com.teambrella.android.ui.base.TeambrellaDataHostActivity;
import com.teambrella.android.ui.teammate.TeammateActivity;
import com.teambrella.android.util.StatisticHelper;

import io.reactivex.Notification;
import io.reactivex.disposables.Disposable;
import jp.wasabeef.picasso.transformations.CropCircleTransformation;

/**
 * Claim Activity
 */
public class ClaimActivity extends TeambrellaDataHostActivity implements IClaimActivity {

    private static final String CLAIM_DATA_TAG = "claim_data_tag";
    private static final String VOTE_DATA_TAG = "vote_data_tag";
    private static final String UI_TAG = "ui";
    private static final String EXTRA_URI = "uri";
    private static final String EXTRA_MODEL = "model";
    private static final String EXTRA_TEAM_ID = "team_id";
    private static final String EXTRA_CLAIM_ID = "claimId";

    private static final int DEFAULT_REQUEST_CODE = 4;


    private Disposable mDisposal;


    private int mClaimId;
    private int mTeamId;
    private TextView mTitle;
    private TextView mSubtitle;
    private ImageView mIcon;

    private Snackbar mSnackBar;


    public static Intent getLaunchIntent(Context context, int id, String model, int teamId) {
        return new Intent(context, ClaimActivity.class)
                .putExtra(EXTRA_URI, TeambrellaUris.getClaimUri(id))
                .putExtra(EXTRA_MODEL, model)
                .putExtra(EXTRA_CLAIM_ID, id)
                .putExtra(EXTRA_TEAM_ID, teamId);
    }

    public static void start(Context context, int id, String model, int teamId) {
        context.startActivity(getLaunchIntent(context, id, model, teamId));
    }


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        final Intent intent = getIntent();
        mClaimId = intent.getIntExtra(EXTRA_CLAIM_ID, -1);
        mTeamId = intent.getIntExtra(EXTRA_TEAM_ID, -1);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_claim);
        FragmentManager fragmentManager = getSupportFragmentManager();
        if (fragmentManager.findFragmentByTag(UI_TAG) == null) {
            fragmentManager.beginTransaction()
                    .add(R.id.container, ADataProgressFragment.getInstance(new String[]{CLAIM_DATA_TAG, VOTE_DATA_TAG}, ClaimFragment.class), UI_TAG)
                    .commit();
        }

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeAsUpIndicator(R.drawable.ic_arrow_back_vector);
            actionBar.setDisplayOptions(actionBar.getDisplayOptions() | ActionBar.DISPLAY_SHOW_CUSTOM);
            actionBar.setCustomView(R.layout.claim_toolbar_view);
            View view = actionBar.getCustomView();
            mTitle = view.findViewById(R.id.title);
            mSubtitle = view.findViewById(R.id.subtitle);
            mIcon = view.findViewById(R.id.icon);
            Toolbar parent = (Toolbar) view.getParent();
            parent.setPadding(0, 0, 0, 0);
            parent.setContentInsetsAbsolute(0, 0);
        }
        setTitle(getIntent().getStringExtra(EXTRA_MODEL));
    }


    @Override
    protected void onStart() {
        super.onStart();
        mDisposal = getObservable(CLAIM_DATA_TAG).subscribe(this::onDataUpdated);
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (mDisposal != null && !mDisposal.isDisposed()) {
            mDisposal.dispose();
        }
    }

    @Override
    protected String[] getDataTags() {
        return new String[]{CLAIM_DATA_TAG, VOTE_DATA_TAG};
    }

    @Override
    protected TeambrellaDataFragment getDataFragment(String tag) {
        switch (tag) {
            case CLAIM_DATA_TAG:
                return TeambrellaDataFragment.getInstance(getIntent().getParcelableExtra(EXTRA_URI));
            case VOTE_DATA_TAG:
                return TeambrellaDataFragment.getInstance(null);
        }

        return null;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected String[] getPagerTags() {
        return new String[]{};
    }

    @Override
    protected TeambrellaDataPagerFragment getDataPagerFragment(String tag) {
        return null;
    }

    @Override
    public void setTitle(String title) {
        if (mTitle != null) {
            mTitle.setText(title);
        }
    }

    @Override
    public void setSubtitle(String subtitle) {
        if (mSubtitle != null) {
            mSubtitle.setText(subtitle);
        }
    }


    @Override
    public void postVote(int vote) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        TeambrellaDataFragment dataFragment = (TeambrellaDataFragment) fragmentManager.findFragmentByTag(VOTE_DATA_TAG);
        if (dataFragment != null) {
            dataFragment.load(TeambrellaUris.getClaimVoteUri(mClaimId, vote));
        }
        StatisticHelper.onClaimVote(getTeamId(), vote);
    }


    @Override
    public void showSnackBar(@StringRes int text) {
        if (mSnackBar == null) {
            mSnackBar = Snackbar.make(findViewById(R.id.container), text, Snackbar.LENGTH_LONG);

            mSnackBar.addCallback(new Snackbar.Callback() {
                @Override
                public void onShown(Snackbar sb) {
                    super.onShown(sb);
                }

                @Override
                public void onDismissed(Snackbar transientBottomBar, int event) {
                    super.onDismissed(transientBottomBar, event);
                    mSnackBar = null;
                }
            });

            mSnackBar.show();
        }
    }

    @Override
    public int getClaimId() {
        return mClaimId;
    }

    @Override
    public int getTeamId() {
        return mTeamId;
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        load(CLAIM_DATA_TAG);
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void launchActivity(Intent intent) {
        startActivityForResult(intent, DEFAULT_REQUEST_CODE);
    }

    protected void onDataUpdated(Notification<JsonObject> notification) {
        if (notification.isOnNext()) {
            JsonWrapper response = new JsonWrapper(notification.getValue());
            JsonWrapper data = response.getObject(TeambrellaModel.ATTR_DATA);
            JsonWrapper basic = data.getObject(TeambrellaModel.ATTR_DATA_ONE_BASIC);
            if (basic != null) {
                String pictureUri = TeambrellaModel.getImage(TeambrellaServer.BASE_URL, basic.getObject(), TeambrellaModel.ATTR_DATA_AVATAR);
                if (pictureUri != null) {
                    TeambrellaImageLoader.getInstance(this).getPicasso()
                            .load(pictureUri).transform(new CropCircleTransformation()).into(mIcon);
                    mIcon.setOnClickListener(v ->
                            TeammateActivity.start(ClaimActivity.this
                                    , getIntent().getIntExtra(EXTRA_TEAM_ID, 0)
                                    , basic.getString(TeambrellaModel.ATTR_DATA_USER_ID)
                                    , basic.getString(TeambrellaModel.ATTR_DATA_NAME)
                                    , pictureUri));
                }

                setTitle(basic.getString(TeambrellaModel.ATTR_DATA_MODEL));
            }
            mClaimId = data.getInt(TeambrellaModel.ATTR_DATA_ID, 0);

        }
    }
}

File: app/src/main/java/com/teambrella/android/ui/claim/ClaimActivity.kt
package com.teambrella.android.ui.claim

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.google.android.material.snackbar.Snackbar
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.api.server.TeambrellaUris
import com.teambrella.android.services.TeambrellaNotificationServiceClient
import com.teambrella.android.services.push.INotificationMessage
import com.teambrella.android.ui.base.*
import com.teambrella.android.ui.teammate.startTeammateActivity
import com.teambrella.android.ui.util.setAvatar
import com.teambrella.android.util.StatisticHelper
import io.reactivex.Notification

/**
 * Claim Activity
 */
class ClaimActivity : ATeambrellaActivity(), IClaimActivity {

    companion object {

        private const val CLAIM_DATA_TAG = "claim_data_tag"
        private const val VOTE_DATA_TAG = "vote_data_tag"
        private const val UI_TAG = "ui"
        private const val DEFAULT_REQUEST_CODE = 4

        fun getLaunchIntent(context: Context, id: Int, model: String?, teamId: Int): Intent {
            return Intent(context, ClaimActivity::class.java).apply {
                this.teamId = teamId
                this.claimId = id
                this.model = model
                this.uri = TeambrellaUris.getClaimUri(id)
            }
        }

        fun start(context: Context, id: Int, model: String?, teamId: Int) {
            context.startActivity(getLaunchIntent(context, id, model, teamId))
        }
    }


    override var claimId: Int = 0
        get() = if (field > 0) field else intent?.claimId ?: 0

    override var teamId: Int = 0
        get() = if (field > 0) field else intent?.teamId ?: 0


    private var titleView: TextView? = null
    private var subtitleView: TextView? = null
    private var iconView: ImageView? = null
    private var snackBar: Snackbar? = null
    private var topicId: String? = null


    private lateinit var mNotificationClient: ClaimNotificationClient
    private lateinit var mChatBroadCastManager: TeambrellaBroadcastManager

    override val dataTags: Array<String>
        get() = arrayOf(CLAIM_DATA_TAG, VOTE_DATA_TAG)

    override fun onCreate(savedInstanceState: Bundle?) {
        mChatBroadCastManager = TeambrellaBroadcastManager(this)
        mChatBroadCastManager.registerReceiver(mChatBroadCastReceiver)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_claim)

        supportFragmentManager.apply {
            if (findFragmentByTag(UI_TAG) == null) {
                beginTransaction()
                        .add(R.id.container, createDataFragment(arrayOf(CLAIM_DATA_TAG, VOTE_DATA_TAG), ClaimFragment::class.java), UI_TAG)
                        .commit()
            }
        }

        mNotificationClient = ClaimNotificationClient(this)
        mNotificationClient.connect()

        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_arrow_back_vector)
            displayOptions = displayOptions or ActionBar.DISPLAY_SHOW_CUSTOM
            setCustomView(R.layout.claim_toolbar_view)
            customView?.apply {
                titleView = findViewById(R.id.title)
                subtitleView = findViewById(R.id.subtitle)
                iconView = findViewById(R.id.icon)
                (parent as Toolbar).apply {
                    setPadding(0, 0, 0, 0)
                    setContentInsetsAbsolute(0, 0)
                }
            }


        }
        titleView?.text = (intent.model)
        component.inject(this)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        getObservable(CLAIM_DATA_TAG).observe(this, Observer<Notification<JsonObject>> { it?.let { this.onDataUpdated(it) } })
    }


    override fun getDataConfig(tag: String): Bundle? {
        when (tag) {
            CLAIM_DATA_TAG -> return getDataConfig(intent.uri)
            VOTE_DATA_TAG -> return getDataConfig()
        }
        return super.getDataConfig(tag)
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : TeambrellaDataViewModel> getDataViewModelClass(tag: String) = when (tag) {
        CLAIM_DATA_TAG -> ClaimViewModel::class.java as Class<T>
        else -> super.getDataViewModelClass<T>(tag)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun setTitle(title: String) {
        titleView?.text = title
    }

    override fun setSubtitle(subtitle: String) {
        subtitleView?.text = subtitle
    }


    override fun postVote(vote: Int) {
        ViewModelProviders.of(this).get(VOTE_DATA_TAG, TeambrellaDataViewModel::class.java)
                .load(TeambrellaUris.getClaimVoteUri(claimId, vote))
        StatisticHelper.onClaimVote(this, teamId, claimId, vote)
        TeambrellaBroadcastManager(this).notifyClaimVote(claimId)
    }


    override fun showSnackBar(@StringRes text: Int) {
        if (snackBar == null) {
            snackBar = Snackbar.make(findViewById(R.id.container), text, Snackbar.LENGTH_LONG)
            snackBar?.addCallback(object : Snackbar.Callback() {
                override fun onDismissed(transientBottomBar: Snackbar?, event: Int) {
                    super.onDismissed(transientBottomBar, event)
                    snackBar = null
                }
            })
            snackBar?.show()
        }
    }

    override fun onResume() {
        super.onResume()
        mNotificationClient.onResume()
    }


    override fun onPause() {
        super.onPause()
        mNotificationClient.onPause()
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        load(CLAIM_DATA_TAG)
        super.onActivityResult(requestCode, resultCode, data)
    }


    override fun onDestroy() {
        super.onDestroy()
        mNotificationClient.disconnect()
        mChatBroadCastManager.unregisterReceiver(mChatBroadCastReceiver)
    }

    override fun launchActivity(intent: Intent) {
        startActivityForResult(intent, DEFAULT_REQUEST_CODE)
    }

    private fun onDataUpdated(notification: Notification<JsonObject>) {
        notification.takeIf { it.isOnNext }?.value?.data?.let { _data ->
            _data.basic?.let { _basic ->
                val pictureUri = _basic.avatar
                iconView?.setAvatar(getImageLoader()?.getImageUrl(pictureUri))
                iconView?.setOnClickListener {
                    startTeammateActivity(this@ClaimActivity, intent.teamId
                            ?: 0, _basic.userId!!, _basic.name, pictureUri)
                }
                titleView?.text = _basic.model
            }

            topicId = _data.discussion?.topicId
            claimId = _data.intId ?: 0
        }
    }

    private fun markTopicRead(topicId: String?) {
        ViewModelProviders.of(this).get(CLAIM_DATA_TAG, ClaimViewModel::class.java).markTopicRead(topicId!!)
    }

    private inner class ClaimNotificationClient internal constructor(context: Context) : TeambrellaNotificationServiceClient(context) {

        private var mResumed: Boolean = false
        private var mUpdateOnResume: Boolean = false

        override fun onPushMessage(message: INotificationMessage): Boolean {
            val topicId = message.topicId
            if (topicId != null && topicId == this@ClaimActivity.topicId) {
                if (mResumed) {
                    load(CLAIM_DATA_TAG)
                } else {
                    mUpdateOnResume = true
                }
            }
            return false
        }

        internal fun onPause() {
            mResumed = false
        }

        internal fun onResume() {
            mResumed = true
            if (mUpdateOnResume) {
                load(CLAIM_DATA_TAG)
                mUpdateOnResume = false
            }
        }
    }

    private val mChatBroadCastReceiver = object : TeambrellaBroadcastReceiver() {
        override fun onTopicRead(topicId: String) {
            super.onTopicRead(topicId)
            if (topicId == this@ClaimActivity.topicId) {
                markTopicRead(this@ClaimActivity.topicId)
            }
        }
    }
}
